src/braket/default_simulator/openqasm/data_manipulation.py

```diff
@@ -0,0 +1,656 @@
+def convert_range_def_to_slice(range_def: RangeDefinition) -> slice:
+    """Convert AST node into Python slice object"""
+    buffer = np.sign(range_def.step.value) if range_def.step is not None else 1
+    start = range_def.start.value if range_def.start is not None else None
+    stop = (
+        range_def.end.value + buffer
+        if not (range_def.end is None or range_def.end.value == -1)
+        else None
+    )
+    step = range_def.step.value if range_def.step is not None else None
+    return slice(start, stop, step)
+
+
+def convert_range_def_to_range(range_def: RangeDefinition) -> range:
+    """Convert AST node into Python range object"""
+    buffer = np.sign(range_def.step.value) if range_def.step is not None else 1
+    start = range_def.start.value if range_def.start is not None else 0
+    stop = range_def.end.value + buffer
+    step = range_def.step.value if range_def.step is not None else 1
+    return range(start, stop, step)
+
+
+def convert_discrete_set_to_list(discrete_set: DiscreteSet) -> list:
+    """Convert AST node into Python list object"""
+    return [x.value for x in discrete_set.values]
+
+
+@singledispatch
+def get_elements(
+    value: ArrayLiteral, index: IndexElement, type_width: Optional[int] = None
+) -> ArrayLiteral:
+    """Get elements of an Array, given an index."""
+    if isinstance(index, DiscreteSet):
+        return ArrayLiteral([get_elements(value, [i]) for i in index.values])
+    first_index = convert_index(index[0])
+    if isinstance(first_index, int):
+        if not index[1:]:
+            return value.values[first_index]
+        return get_elements(value.values[first_index], index[1:], type_width)
+    # first_index is a slice
+    index_as_range = range(len(value.values))[first_index]
+    if not index[1:]:
+        return ArrayLiteral([value.values[ix] for ix in index_as_range])
+    return ArrayLiteral(
+        [get_elements(value.values[ix], index[1:], type_width) for ix in index_as_range]
+    )
+
+
+@get_elements.register
+def _(value: IntegerLiteral, index: IndexElement, type_width: int) -> ArrayLiteral:
+    """Get elements of an integer's boolean representation, given an index"""
+    binary_rep = ArrayLiteral(
+        [BooleanLiteral(x == "1") for x in np.binary_repr(value.value, type_width)]
+    )
+    return get_elements(binary_rep, index)
+
+
+def create_empty_array(dims: List[IntegerLiteral]) -> ArrayLiteral:
+    """Create an empty Array of given dimensions"""
+    if len(dims) == 1:
+        return ArrayLiteral([None] * dims[0].value)
+    return ArrayLiteral([create_empty_array(dims[1:])] * dims[0].value)
+
+
+def convert_index(index: Union[RangeDefinition, IntegerLiteral]) -> Union[int, slice]:
+    """Convert unspecified index type to Python object"""
+    if isinstance(index, RangeDefinition):
+        return convert_range_def_to_slice(index)
+    else:  # IntegerLiteral:
+        return index.value
+
+
+def flatten_indices(indices: List[IndexElement]) -> list:
```

Comment:
Most of the functionality here is tested through the interpreter. ie instead of
```
assert flatten_indices([[1, 2], [3], [4, 5]]) == [1, 2, 3, 4, 5]
```
The test would be
```
qasm = """
array[int[8], 6, 6, 6, 6, 6] arr = {...};
int[8] get_1 = arr[1, 2][3][4, 5];
int[8] get_2 = arr[1, 2, 3, 4, 5];
int[8] get_3 = arr[1][2][3][4][5];
"""
# run program
# assert get_1 == get_2 == get_3
```

Maybe we can chat about testing strategy here, I'd love to hear your thoughts

Reply:
Yeah, but then I wonder why it's not just a private function in the interpreter? 

It's a little easier to see the test coverage when you're testing the function directly. Ie. 
Are we testing all the boundary cases ([], [1, [2, 3]][4][][5][None][6]), etc

Reply:
Yeah, I don't see any reason to just have it private in program_context; in general, we should make things private as much as we can, and only the core API public. This applies to many of the utility methods in this file.